### PR TITLE
add feature: symlink and hardlink options

### PR DESCRIPTION
**Pull request by evanrichter**
I made a new PR since my first was messy with commits, and my second try is now outdated.

This would close issue #7 

The added options are `-s <dest dir>` for symlinks and `-H <dest dir>` for hardlinks.
Symlinks only support Unix based OSes for simplicity (windows will exit with an error message)
Hardlinks work on Unix and windows platforms

The relevant help text looks like:
```
imdb-rename 0.1.5
Andrew Gallant <jamslam@gmail.com>

USAGE:
    imdb-rename [OPTIONS] [file]...

OPTIONS:
                              ...
    -H, --hardlink <folder>          Place hardlink in <folder> instead of moving.
                              ...
    -s, --symlink <folder>           Place symlink in <folder> instead of moving. Be careful of
                                     absolute vs. relative links: the target of the symlink will
                                     match the given input. (Unix only feature)
                              ...

ARGS:
    <file>...    One or more files to rename.
```

### CODE REVIEW

**Comment by evanrichter** [comment]
@BurntSushi would it be possible to merge this in or leave a comment for improvement? I'm sure you're busy but I would appreciate a review since you said this feature would be welcome.

File: src/main.rs [hunk 1]
```diff
@@ -16,6 +16,7 @@ use crate::util::{choose, read_yesno, write_tsv};
 mod download;
 mod logger;
 mod rename;
+use rename::RenameAction;
```
Comment (BurntSushi) [inline src/main.rs]: Could you please:

1. Put this `use` up with the `use::crate` statements above? In fact, please just merge it with `use crate::rename::RenameBuilder;`.
2. Please import `RenameAction` unqualified to avoid the stutter when writing out `rename::RenameAction`.

File: src/main.rs [hunk 2]
```diff
@@ -167,8 +176,25 @@ impl Args {
...
             .parse()?;
+        let (rename_action, dest_dir): (RenameAction, Option<PathBuf>) =
+            if matches.is_present("symlink") {
+                if cfg!(windows) {
+                    failure::bail!(
+                        "symlink not supported on Windows, try hardlink");
+                }
+                (RenameAction::Symlink, Some(matches.value_of_os("symlink")
+                                             .unwrap()
+                                             .to_os_string().into()))
```
Comment (BurntSushi) [inline src/main.rs]: Could you please format the code in a style consistent with the surrounding code? e.g.,

```
(RenameAction::Symlink, Some(
    matches.value_of_os("symlink")
        .unwrap().to_os_string().into()
))
```

File: src/main.rs [hunk 3]
```diff
@@ -167,8 +176,25 @@ impl Args {
...
+            } else if matches.is_present("hardlink") {
+                (RenameAction::Hardlink, Some(matches.value_of_os("hardlink")
+                                              .unwrap()
+                                              .to_os_string().into()))
```
Comment (BurntSushi) [inline src/main.rs]: Do you specifically have a use case for hardlinks? If not, please remove it. Hardlinks are weird and I'd rather not encourage their use.

File: src/rename.rs [hunk 4]
```diff
@@ -35,14 +57,37 @@ impl RenameProposal {
...
+                    use std::os::unix;
+                    unix::fs::symlink(&self.src, &self.dst).map_err(|e| {
+                        format_err!(
+                            "error symlinking '{}' to '{}': {}",
+                            self.src.display(),
+                            self.dst.display(),
+                            e,
+                        )
+                    })?;
+                }
```
Comment (BurntSushi) [inline src/rename.rs]: Could you return an error for `#[cfg(not(unix))]`? I know this path isn't exercised on `not(unix)`, but that might not always be true in the future. Otherwise, this would just silently return success without actually doing anything.

File: src/main.rs [hunk 5]
```diff
@@ -302,6 +329,20 @@ fn app() -> clap::App<'static, 'static> {
...
              .long("update-index")
              .help("Forcefully re-indexes the IMDb data and then exits."))
+        .arg(Arg::with_name("symlink")
+             .long("symlink")
+             .short("s")
+             .value_name("folder")
+             .conflicts_with("hardlink")
+             .help("Place symlink in <folder> instead of moving. Be careful of \
+                   absolute vs. relative links: the target of the symlink will \
+                   match the given input. (Unix only feature)"))
```
Comment (BurntSushi) [inline src/main.rs]: I'm having trouble figuring out why `symlink` takes an argument here. Wouldn't a destination directory be more appropriate as a separate standalone flag? It seems like an orthogonal concept to symlinks...

Comment (evanrichter) [reply src/main.rs:339]: I agree. I'll add a `--destination` / `-d` Arg. this would also apply to normal renames. the file would be renamed, and moved to the destination (only if specified of course).
in the situation `symlink` is specified but no destination is specified, a symlinked file will appear next to the original file in the same directory.

File: src/main.rs (hunk 3 above)
Comment (evanrichter) [reply src/main.rs:191]: I currently use symlinks because my fuse filesystem doesn't support hardlinks, otherwise I would use hardlinks.

I agree they can be weird, but I think the user should decide the best type of file link for their situation

situation: I want a well organized media directory with sane filenames. Other applications still need access to the original filenames. with hardlinks, there is a cleaner separation between the two uses. If I delete from the media folder, it will still be available for the other application, and vice versa. When neither hardlink exists, the file is deleted.

With symlinks, I only get one side of that indirection. I can't delete from the original location and keep the media folder intact. The file is deleted at once.

File: src/main.rs (hunk 2 above)
Comment (evanrichter) [reply src/main.rs:187]: of course!

File: src/main.rs (hunk 1 above)
Comment (evanrichter) [reply src/main.rs:19]: :+1:

File: src/rename.rs (hunk 4 above)
Comment (evanrichter) [reply src/rename.rs:81]: :+1:

File: src/main.rs (hunk 5 above)
Comment (evanrichter) [reply src/main.rs:339]: @BurntSushi because of absolute/relative paths concerning symlinks, these are the behaviors I imagine being useful defaults:

| Action | no destination specified | `-d outdir` specified |
| --- | --- | --- |
| rename | file is renamed in place (current behavior) | file is moved to `outdir` after being renamed. In code this would still be a single `std::fs::rename` |
| symlink | well-named link appears in current directory and points to the original filepath, however given.<br><br>If the original path is relative (`imdb-rename -s ./movies/Movie.2010.1080p.mkv`) then the link will be relative and only work if the symlink remains in the current directory: `Movie (2010).mkv -> ./movies/Movie.2010.1080p.mkv`.<br><br>If the path given is absolute (`imdb-rename -s /home/evan/movies/Movie.2010.1080p.mkv`) then the link will be absolute, and will work as expected even if moved from the current location. | well-named link appears in the folder `outdir` as requested, and will point to the absolute path of the files given, no matter how the files were specified on the commandline. This is the only way that the created symlink would know how to actually find the file in question without a lot more work, and is probably what the user expects anyway. |
| hardlink | named hardlink appears in current directory, and links correctly to original file regardless of filepath (just because of how hardlinks work) | well-named hardlink appears in `outdir`, and links correctly to original file regardless of filepath (just because of how hardlinks work) |

Comment (evanrichter) [reply src/main.rs]: alright this matrix as is, is implemented in the pr now

**Comment by evanrichter** [comment]
thanks for the comments and being patient with this pr :)
